sdk/src/main/scala/com/horizen/account/state/AccountStateView.scala

```diff
@@ -103,44 +98,111 @@ class AccountStateView(private val metadataStorageView: AccountStateMetadataStor
           val recipientProposition = new AddressProposition(BytesUtils.reverseBytes(ftOut.propositionBytes.take(com.horizen.account.utils.Account.ADDRESS_SIZE)))
 
           // stateDb will implicitly create account if not existing yet
-          view.addBalance(recipientProposition.address(), value)
+          addBalance(recipientProposition.address(), value)
           log.debug(s"added FT amount = $value to address=$recipientProposition")
       }
     })
   }
 
-  def getOrderedForgingStakeInfoSeq : Seq[ForgingStakeInfo] = {
-
+  def getOrderedForgingStakeInfoSeq: Seq[ForgingStakeInfo] = {
     val forgerStakeList = forgerStakesProvider.getListOfForgers(this)
 
     forgerStakeList.map {
-      item => ForgingStakeInfo(
-        item.forgerStakeData.forgerPublicKeys.blockSignPublicKey,
-        item.forgerStakeData.forgerPublicKeys.vrfPublicKey,
-        ZenWeiConverter.convertWeiToZennies(item.forgerStakeData.stakedAmount))
+      item =>
+        ForgingStakeInfo(
+          item.forgerStakeData.forgerPublicKeys.blockSignPublicKey,
+          item.forgerStakeData.forgerPublicKeys.vrfPublicKey,
+          ZenWeiConverter.convertWeiToZennies(item.forgerStakeData.stakedAmount))
     }.sorted(Ordering[ForgingStakeInfo].reverse)
   }
 
 
   def setupTxContext(tx: EthereumTransaction): Unit = {
-    // TODO
+    // TODO: set context for the created events/logs assignment
   }
 
-  override def applyTransaction(tx: SidechainTypes#SCAT): Try[AccountStateView] = Try {
-    if (tx.isInstanceOf[EthereumTransaction]) {
-      val ethTx = tx.asInstanceOf[EthereumTransaction]
-      setupTxContext(ethTx)
-      val message: Message = Message.fromTransaction(ethTx)
-      val processor = messageProcessors.find(_.canProcess(message, this)).getOrElse(
-        throw new IllegalArgumentException(s"Transaction ${ethTx.id} has no known processor.")
-      )
-      processor.process(message, this) match {
-        case success: ExecutionSucceeded => this // TODO
-        case failed: ExecutionFailed => this // TODO
-        case invalid: InvalidMessage => throw new Exception(s"Transaction ${ethTx.id} is invalid.", invalid.getReason)
-      }
-    } else
+  private def preCheck(tx: EthereumTransaction): BigInteger = {
+    // We are sure that transaction is semantically valid (so all the tx fields are valid)
+    // and was successfully verified by ChainIdBlockSemanticValidator
+
+    // Check signature
+    if (!tx.getSignature.isValid(tx.getFrom, tx.messageToSign()))
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: signature is invalid")
+
+    // Check that "from" is EOA address
+    if(!isEoaAccount(tx.getFrom.address()))
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: from account is not EOA")
+
+    // Check the nonce
+    val stateNonce: BigInteger = getNonce(tx.getFrom.address())
+    val txNonce: BigInteger = tx.getNonce
+    stateNonce.compareTo(txNonce) match {
```

Comment:
Apparently this match has problems and throws a scala.MatchError: 0 (of class java.lang.Integer), I tried using:

```
    val result = stateNonce.compareTo(txNonce)
    if (result > 0) {
      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: nonce ${txNonce} is to high (expected nonce is $stateNonce)")
    } else if (result < 0) {
      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: nonce ${txNonce} is to low (expected nonce is $stateNonce)")
    }
```